File: web/src/app/core/services/notificaciones.service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { ApiService } from './api.service';
import { Notificacion, NotificacionApi, mapNotificacionApiToNotificacion } from '../models/notificacion.model';

@Injectable({
  providedIn: 'root'
})
export class NotificacionesService {
  private readonly apiService = inject(ApiService);

  readonly notificaciones = signal<Notificacion[]>([]);
  readonly contadorNoLeidas = signal<number>(0);
  readonly isLoading = signal<boolean>(false);

  readonly hayNoLeidas = computed(() => this.contadorNoLeidas() > 0);

  /**
   * Carga las últimas notificaciones del usuario (GET /api/notificaciones)
   */
  async cargarNotificaciones(): Promise<void> {
    this.isLoading.set(true);
    try {
      const response = await firstValueFrom(
        this.apiService.get<NotificacionApi[]>('/api/notificaciones', undefined, undefined, 'usuarios')
      );
      this.notificaciones.set((response || []).map(mapNotificacionApiToNotificacion));
    } catch (err) {
      console.error('[NotificacionesService] Error al cargar notificaciones:', err);
    } finally {
      this.isLoading.set(false);
    }
  }

  /**
   * Carga el contador de notificaciones no leídas (GET /api/notificaciones/contador-no-leidas)
   */
  async cargarContador(): Promise<void> {
    try {
      const response = await firstValueFrom(
        this.apiService.get<{ count: number }>('/api/notificaciones/contador-no-leidas', undefined, undefined, 'usuarios')
      );
      this.contadorNoLeidas.set(response?.count ?? 0);
    } catch (err) {
      console.error('[NotificacionesService] Error al cargar contador de no leídas:', err);
    }
  }

  /**
   * Marca una notificación individual como leída (PATCH /api/notificaciones/{id}/leer)
   * Actualización optimista: refleja el cambio de inmediato en la UI.
   */
  async marcarComoLeida(id: string): Promise<void> {
    const yaEstabaLeida = this.notificaciones().find(n => n.id === id)?.leida;
    this.notificaciones.update(lista => lista.map(n => n.id === id ? { ...n, leida: true } : n));
    if (!yaEstabaLeida) {
      this.contadorNoLeidas.update(c => Math.max(0, c - 1));
    }

    try {
      await firstValueFrom(
        this.apiService.patch<void>(`/api/notificaciones/${id}/leer`, {}, undefined, 'usuarios')
      );
    } catch (err) {
      console.error('[NotificacionesService] Error al marcar notificación como leída:', err);
    }
  }

  /**
   * Marca todas las notificaciones del usuario como leídas (POST /api/notificaciones/marcar-todas)
   * Actualización optimista: refleja el cambio de inmediato en la UI.
   */
  async marcarTodasComoLeidas(): Promise<void> {
    this.notificaciones.update(lista => lista.map(n => ({ ...n, leida: true })));
    this.contadorNoLeidas.set(0);

    try {
      await firstValueFrom(
        this.apiService.post<void>('/api/notificaciones/marcar-todas', {}, undefined, 'usuarios')
      );
    } catch (err) {
      console.error('[NotificacionesService] Error al marcar todas las notificaciones como leídas:', err);
    }
  }
}

File: mobile/lib/Services/notificaciones_service.dart
import 'dart:convert';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import '../Models/notificacion.dart';
import 'app_controller.dart';

class NotificacionesService {
  final AppController controller;

  NotificacionesService(this.controller);

  String get baseUrl =>
      dotenv.env['API_BASE_URL_USUARIOS'] ??
      'https://usuarios-api-n1qi.onrender.com';

  Future<Map<String, String>> _getHeaders() async {
    final token = await controller.getValidAccessToken();
    return {
      if (token != null && token.isNotEmpty) 'Authorization': 'Bearer $token',
      'Content-Type': 'application/json',
    };
  }

  /// Carga la lista de notificaciones del usuario (GET /api/notificaciones)
  Future<List<Notificacion>> getNotificaciones() async {
    try {
      final url = '$baseUrl/api/notificaciones';
      final response = await controller.httpClient.get(
        Uri.parse(url),
        headers: await _getHeaders(),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        if (response.body.isEmpty) return [];
        final decoded = jsonDecode(response.body);
        if (decoded is List) {
          return decoded
              .map((item) => Notificacion.fromJson(item as Map<String, dynamic>))
              .toList();
        }
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  /// Carga el conteo de notificaciones no leídas (GET /api/notificaciones/contador-no-leidas)
  Future<int> getContadorNoLeidas() async {
    try {
      final url = '$baseUrl/api/notificaciones/contador-no-leidas';
      final response = await controller.httpClient.get(
        Uri.parse(url),
        headers: await _getHeaders(),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        if (response.body.isEmpty) return 0;
        final decoded = jsonDecode(response.body);
        if (decoded is Map<String, dynamic>) {
          return (decoded['count'] as num?)?.toInt() ?? 0;
        }
      }
      return 0;
    } catch (e) {
      return 0;
    }
  }

  /// Marca una notificación como leída (PATCH /api/notificaciones/{id}/leer)
  Future<bool> marcarComoLeida(String id) async {
    try {
      final url = '$baseUrl/api/notificaciones/$id/leer';
      final response = await controller.httpClient.patch(
        Uri.parse(url),
        headers: await _getHeaders(),
        body: jsonEncode({}),
      );

      return response.statusCode >= 200 && response.statusCode < 300;
    } catch (e) {
      return false;
    }
  }

  /// Marca todas las notificaciones como leídas (POST /api/notificaciones/marcar-todas)
  Future<bool> marcarTodasComoLeidas() async {
    try {
      final url = '$baseUrl/api/notificaciones/marcar-todas';
      final response = await controller.httpClient.post(
        Uri.parse(url),
        headers: await _getHeaders(),
        body: jsonEncode({}),
      );

      return response.statusCode >= 200 && response.statusCode < 300;
    } catch (e) {
      return false;
    }
  }
}
